### ISSUE TITLE
`npm install` should deduplicate its tree as part of the install process

### ISSUE THREAD

**Issue by iarna** [npm/npm#6912]
This is intended to solve a few larger project goals:

1. Installs should be smaller, without having to explicitly ask for it.
2. For Windows user's sake, we should do our best to avoid long path names. Many windows tools are limited to 256 characters.

As such, @othiym23 and I have decided this should be a maximally flat deduplication, which means if you have a dep tree of:

    A → B → C

Then if you install A, you'll get B and C next to it in your node modules. That is, it'll be installed as:

    A
    B
    C

But we'll hold onto metadata to know why B and C were installed such that removing A will remove B and C as it would now.

As for devDependencies, they shouldn't have their dependencies folded in as above. Instead each of them should essentially act as its own root. So if A above were a dev dependency then it would be installed as:

    A → B
        C

The following work remains to be done:

* [x] The core deduplicated install work needs to be completed. It is currently in progress.
* #7273 `npm outdated` needs to structure its output in a way that makes sense with a flat node_modules folder. 
* #7272 `npm ls` needs to structure its output in a way that makes sense with a flat node_modules folder. 

See also: #4761, #4037, #5465

**Comment by timoxley** [comment]
Wholly +1 on this. 

Note that this is going to break a lot more package.json scripts incorrectly relying on executables being installed in `node_modules/.bin`. I'm wondering if perhaps it's worth putting an an explicit check for `node_modules/.bin` in scripts and shooting off a warning, because there's few valid reasons to reference `node_modules/.bin` directly in a package.json.

> there's a risk they'll come to depend on things that aren't actually part of the install. 

We could adapt some existing require detecting tooling to detect this pretty easily.

**Comment by isaacs** [comment]
> I'm wondering if perhaps it's worth putting an an explicit check for node_modules/.bin in scripts and shooting off a warning, because there's few valid reasons to reference node_modules/.bin directly in package.json scripts, or anywhere for that matter.

That's really smart.  We should just strip `node_modules/.bin` from the start of them, I think, and maybe also warn.

**Comment by iarna** [comment]
We actually *already do* strip `node_modules/.bin` already in `read-package-json`, here:

https://github.com/npm/read-package-json/commit/56c7f5de

I'm going to be moving this over to `normalize-package-data` and adding a warn.

**Comment by piranna** [comment]
I've just re-read the initial post. Why ```devDependencies``` need to have a diferent behaviour to regular ```dependencies```? Why they can be also installed in a flat structure? This way they would reuse some of the dependencies for themselves since they are also managed by versioning...

**Comment by iarna** [comment]
TLDR: The actual implementation of this ended up not needing `devDependencies` to be treated specially.

My previous concerns around `devDependencies` came about when I was envisioning a more aggressive deduping process that would downgrade (within semver constraints) modules in order to allow greater deduplication. This would allow for:

```
dep: A → B ^1.2
devdep: C → B ~1.2
```
And 1.3 of B is available

At which point checking out the repo and running `npm install` would get you:
```
A
C
B@1.2
```

Versus installing the module with `npm install --production` (or installing it as a dep of another module):
```
A
B@1.3
```

Which would mean that the developer would be testing against a different version than any of their users. Yes, semver says that should be ok, but when it wasn't it would surely cause confusion.

This is why I was discussing keeping dev dependencies separate– so that they wouldn't result in a different version of modules than `npm install --production`.

Now, all that said, I ended up opting for not deduplicating that aggressively and as such, this issue won't turn up, even with dev dependencies flattened into the rest of your node_modules.

**Comment by piranna** [comment]
Ok, thanks for the info :-)

**Comment by dmail** [comment]
I avoid npm because it currently duplicates the dependencies, so a big big big +1

**Comment by KidkArolis** [comment]
So, just to double check, are you planning to dedupe devDependencies the same way as dependencies? 

I think it's important to do so. E.g. if my project depends on jquery or react, but it also *indirectly* depends on jquery or react in devDependencies, deduping devDependencies is a desired behaviour so that subtle issues with running multiple copies of jquery/react in your tests are avoided. In general, I don't see any reason not to dedupe devDependencies (other than your earlier concerns about the downgrading strategy).

The reason I'm double checking is that in the current implementation devDeps are not being deduped: https://github.com/npm/npm/blob/07350a8422f2e876511fc5fe16b98852e97d84ed/lib/dedupe.js#L344-L347.

**Comment by piranna** [comment]
> So, just to double check, are you planning to dedupe devDependencies the same way as dependencies?

According to @iarna, "TLDR: The actual implementation of this ended up not needing devDependencies to be treated specially". So yes, ```devDependencies``` will be deduped, and that are great news because I have the same problem with QUnit and [qunit-cli](https://github.com/devongovett/qunit-cli) :-)

By the way, dedupe would allow to ```require()``` packages not defined in ```package.json```, that could lead to bad defined dependencies... Is this the expected behaviour, or is intended that Node.js ```require()``` only allow to use "logic" dependencies (defined in ```package.json```), or it will still allow to use "phisical" dependencies (everything that's stored on ```node_modules``` folder)?

**Comment by iarna** [comment]
@piranna Well, seeing as npm has no ability to control the node module loader, no, that behavior is not changing. If you want THAT behavior to change, you'd have to convince the node.js and/or io.js teams to change it. As has always been the case with deduplication, modules hoisted to the top level can be required, even if they aren't in your package.json.

**Comment by piranna** [comment]
Yeah, I supossed that, ```package.json``` is mostly focused on ```npm``` and as you pointed you can require a locally manually installed package, so seems there's no easy solution... :-/ Probably if they could read the ```package.json``` file they could show a warning when requiring packages not defined there...

**Comment by piranna** [comment]
[I've open an issue asking for this feature](https://github.com/joyent/node/issues/9354), probably it will not be included, but who knows...

**Comment by mzgol** [comment]
Could you open one for io.js as well? You have a higher chance of someone
actually replying there. :P

-- 
Michał Gołębiowski

**Comment by piranna** [comment]
> Could you open one for io.js as well? You have a higher chance of someone
> actually replying there. :P

[Done](https://github.com/iojs/io.js/issues/1091) ;-)

**Comment by piranna** [comment]
On io.js [they think](https://github.com/iojs/io.js/issues/1091#issuecomment-78171284) it would be better to check the dependencies in package.json and show the warning when doing ```npm publish```. That's not a bad idea, too...

**Comment by iarna** [comment]
I'm confused as to what you think `npm publish` would be warning about? Are you suggesting that `npm publish` should try to statically analyze all your source code to see if you require something that's not a dependency? If so, I would encourage you to write this as a tool that can be installed as a `prepublish` hook.

**Comment by piranna** [comment]
> Are you suggesting that npm publish should try to statically analyze all your source code to see if you require something that's not a dependency?

Not me but at io.js, but yes :-)


> I would encourage you to write this as a tool that can be installed as a prepublish hook.

That's another option, too... Didn't though about the npm hooks...

**Comment by parshap** [comment]
> Are you suggesting that npm publish should try to statically analyze all your source code to see if you require something that's not a dependency? If so, I would encourage you to write this as a tool that can be installed as a prepublish hook.

[dependency-check][] does this.

[dependency-check]: https://www.npmjs.com/package/dependency-check

**Comment by piranna** [comment]
Yeah, dependency-check does what I'm talking about, problem with this thing is that's and opt-in instead of being the default behaviour, so it will only check the dependencies if the developer has it installed.

**Comment by iarna** [comment]
Closing this as fixed in [npm/npm#multi-stage](https://github.com/npm/npm/tree/multi-stage)

**Comment by piranna** [comment]
Great! :-D Could you stimate when/in what version this could be integrated in master? Maybe npm 3?

**Comment by jmm** [comment]
@iarna is there any plan to do automatic deduping as part of `npm update`? I'm trying to figure out the best way to handle this scenario:

```json
{
  "name" : "A",
  "dependencies": {
    "B": "^1.2.3",
    "C": "^4.5.6",
}
```

Pinned `C`:
```json
{
  "name" : "B",
  "dependencies": {
    "C": "4.5.6"
}
```

It's desirable for `A` to depend on the same version of `C` that `B` does without having to manually keep them in sync by updating a pinned version number in `A` every time it changes in `B`.

**Comment by ivan-kleshnin** [comment]
This was a huge mistake I should say. Say, you had a `lodash.isarray` dependency, installed a lot of `gulp` packages, removed `lodash.isarray` now all those `gulp` packages is broken because `isarray` was deduped and now you need either to reinstall everything or to manually discover and reinstall broken things.

**Comment by FractalizeR** [comment]
>I have a `lodash.isarray` dependency, install a lot of gulp packages, remove `lodash.isarray` dependency...

Well, I guess we should not allow a required dependency to be uninstalled without the use of something like `--force` key. This is a part of dependency management.

**Comment by piranna** [comment]
Or remove the other ones recursively, too.

**Comment by FractalizeR** [comment]
@piranna Yep, but some new option should be introduced to do that like `--recursive`

**Comment by piranna** [comment]
> @piranna Yep, but some new option should be introduced to do that like --recursive

:+1:

**Comment by ivan-kleshnin** [comment]
> Well, I guess we should not allow a required dependency to be uninstalled without the use of something like --force key. This is a part of dependency management.

`--force` key sound like you're saying "it's not my problem pal". 
Please explain how do you see "correct" uninstall from the user point of view.
My app is no longer need `lodash.isarray`, so it's clearly NOT a required dependency.
I shouldn't remember all deps of all deps, right? So the problem here is caused exactly by NPM deduplication, not any of my actions.

**Comment by FractalizeR** [comment]
> --force key sounds like you're saying "it's not my problem pal". It does not solve anything.

It solves some cases where computer wants to be smarter than a human. Sometimes I want a computer to do something he thinks dangerous or impossible. For example, rollback migrations on production machine. Such actions must be confirmed with `--force` which means "I know that is dangerous, but I want you to do that no matter what you, brainless machine, think about it".

> Please explain how do you see "correct" uninstall from the user point of view. My app is no longer need lodash.isarray, so it's clearly NOT a required dependency. It's not marked as peerDep either.

Dependency is something, that is required by any module in your system. Not just by the package you requested to be installed. This way it works with every other package manager: yum, apt-get, php-composer etc. I don't see why node should be an exception.

You cannot allow user remove a module for which is known that it is used by other modules. The fact, that user never installed it himself doesn't matter. You must present a user with a clear error message about this fact.

But still, there are some cases when this operation (module installation) MUST be performed disregarding consequences (we are programmers after all, not some silly children). For inspecting aftereffects, for example. Or may be we are going to replace all broken dependency links by a module with the same interface, but didn't edit npm registry. Or may be, module we are removing is used only by another modules, that are going to be pruned by my next command. Or... well, there could be some cases, believe me. Of course user must understand, that the use of `--force` key brings npm registry into a broken state.

> By adding new options you'll just get more tricky cases on the next levels of relations. I'm feeling like NPM is already fairly overcomplicated and we need to remove things, not add.

Here I agree. But I do not consider npm complicated when you compare it to similiar package managers.

**Comment by ivan-kleshnin** [comment]
> You cannot allow user remove a module for which is known that it is used by other modules. The fact, that user never installed it himself doesn't matter. You must present a user with a clear error message about this fact

Ok, i'm trying to remove `lodash.isarray`. I'll get error message. My next steps?
As I surely want to remove `lodash.isarray` I google and discover I should use --force flag. 
I add it. Things got removed. Now I'm in the very same initial situation I've described.
What is improved for my workflow? Nothing, I just got another pesky step.

I can accept flag if it prevents you from doing something *that may be done better*.
But in my case there isn't better way, right? So you just add me one more trouble to deal with and call this *improvement*.

**Comment by FractalizeR** [comment]
> What NPM should do instead is to provide a dialog:
Yes, this is another option. But it is just the same thing - `--force` option just in the `dialog` interactive implementation. Or may be I didn't catch the difference?

But sometimes actions must be done from scripts without user interaction.

**Comment by ivan-kleshnin** [comment]
> But sometimes actions must be done from scripts without user interaction.

Yes. Then just give us **one** flag `-i` to operate in interactive mode in the "smart" helpful way I described. Human memory is far from limitless. It's impossible to remember all flags when they build with *machine-first* approach.

**Comment by FractalizeR** [comment]
Well, `-i` is also a good way :). Edited my post, btw. Incorrectly formatted quotation.

**Comment by parshap** [comment]
> I have a lodash.isarray dependency, install a lot of gulp packages, remove lodash.isarray dependency... now all those gulp packages are broken because isarray was deduped at the moment of their installation.

`$ npm uninstall lodash.isarray` should remove the dependency from the top-level package, but it should not have affect on any other installed dependencies. If *lodash.isarray* was previously deduped because the top-level package shared the that module with a dependency, then `npm uninstall` should move it from the top-level `node_modules/` to the nested dependency's `node_modules/` directory.

(Just my take on how I would expect things to work.)

**Comment by FractalizeR** [comment]
When you do `yum remove` on some package, that will lead to uninstallation of other packages, you get a question (Y/N one) with the list of affected packages unless you provide -y flag, that is treated like a 'yes' answer.

**Comment by ivan-kleshnin** [comment]
> When you do yum remove on some package, that will lead to uninstallation of other packages, you get a question (Y/N one) with the list of affected packages (they will also be uninstalled) unless you provide -y flag, that is treated like a 'yes' answer.

Thank you for this example @FractalizeR. We see that yum is also behaves very unfriendly. 
I don't want to remove other packages just because I remove on of their dep.

This simplest correct solution is what @parshap says. Just reinstall this damned `lodash.isarray` without any dialog for for packages that require it. It should be cached. And even if not – it's better to make one HTTP request without my permission, than provide so awful user experience that NPM does now.

**Comment by pzuraq** [comment]
How will deduplication handle different versions of the same dependency?

```
A → B@1.2
C → B@2.0
```

In this scenario, would we end up with a directory structure like the current `node_modules`?

```
node_modules
├── A
│   └── node_modules
│       └── B
└── C
    └── node_modules
        └── B
```

**Comment by othiym23** [comment]
> In this scenario, would we end up with a directory structure like the current `node_modules` in order to preserve the discovery algorithm of `require`?

Yes. The install / deduplication algorithm will only flatten the dependency tree as much as it can without introducing semver conflicts or other forms of dependency hell. If there is a conflict, then each dependency will get its own version of the conflicting dependency.

**Comment by piranna** [comment]
> If there is a conflict, then each dependency will get its own version of the conflicting dependency.

Will there be a flag on package.json to notify we don't allow conflicts on a package to make sure that there's only one version of it and throw an error if not, similar to the sanity-check done by peerDependencies?

**Comment by othiym23** [comment]
> Will there be a flag on package.json to notify we don't allow conflicts on a package to make sure that there's only one version of it and throw an error if not, similar to the sanity-check done by peerDependencies?

No. There is discussion of a new `browserDependencies` (name not final, almost certain to change, don't hold us to this, etc.) dependency type to to handle dependencies that are guaranteed to be flat and that will error on conflicts, which may or may not install to somewhere other than `node_modules`, but this is not generally desirable behavior for regular npm dependencies, configurably or otherwise.

**Comment by piranna** [comment]
My question was not regarded to browser, but instead about the fact that having duplicated libraries I've found several errors when doing instanceof because the same class was being fetch from two diferent copies of the same package and Node.js understood them as two diferent classes.

**Comment by othiym23** [comment]
That's not an error, and that's why `instanceof` checks are generally considered a bad idea in larger Node apps where a module may not have control over all its dependencies – this is far from the only situation in which Node (or JavaScript in general) will fail instanceof tests. Use duck-typing or branding (attaching a symbol containing its canonical type name as a known property with a special name or Symbol) instead, if you want to check the type of an object. This is not a problem `npm@3` is intended to solve.
